lib/std/Build/Step/WriteFile.zig

```diff
@@ -269,10 +269,37 @@ fn make(step: *Step, prog_node: std.Progress.Node) !void {
         }
     }
     for (write_file.directories.items) |dir| {
-        man.hash.addBytes(dir.source.getPath2(b, step));
+        const full_src_dir_path = dir.source.getPath2(b, step);
+        man.hash.addBytes(full_src_dir_path);
         man.hash.addBytes(dir.sub_path);
         for (dir.options.exclude_extensions) |ext| man.hash.addBytes(ext);
         if (dir.options.include_extensions) |incs| for (incs) |inc| man.hash.addBytes(inc);
+        var src_dir = b.build_root.handle.openDir(full_src_dir_path, .{ .iterate = true }) catch |err| {
+            return step.fail("unable to open source directory '{s}': {s}", .{
+                full_src_dir_path, @errorName(err),
+            });
+        };
+        defer src_dir.close();
+
+        var it = try src_dir.walk(b.allocator);
+        next_entry: while (try it.next()) |entry| {
+            for (dir.options.exclude_extensions) |ext| {
+                if (std.mem.endsWith(u8, entry.path, ext)) continue :next_entry;
+            }
+            if (dir.options.include_extensions) |incs| {
+                for (incs) |inc| {
+                    if (std.mem.endsWith(u8, entry.path, inc)) break;
+                } else {
+                    continue :next_entry;
+                }
+            }
```

Comment:
This logic seems duplicated in https://github.com/ziglang/zig/blob/cc843f95f64bde90ba396517055cab2bd19ce6c8/lib/std/Build/Step/WriteFile.zig#L388-L397 .
We could have a shared function `Directory.Options.isFilePathIncluded(options: Directory.Options, file_path: []const u8) bool` -> `if(!options.isFilePathIncluded(entry)) continue;` to make sure they stay in-sync.